File: arena-rpg.elixir/lib/personagem.ex
defmodule Personagem do
  @derive [Access]
  defstruct [nome: nil, arma: nil, hp: nil, x: nil, y: nil]

  def esta_vivo(this), do: this.hp > 0

  def ataca(this, vitima) do 
    unless esta_armado?(this), do: raise "atacante necessita de uma arma"    
    %{ vitima | hp: vitima.hp - this.arma.dano }
  end

  def esta_armado?(this), do: this.arma || false

  def move_para(this, x, y), do: %{ this | x: x, y: y }
  
  def esta_atingivel?(this, vitima) do 
    unless esta_armado?(this), do: raise "atacante necessita de uma arma"
    distancia_ate(this, vitima) <= this.arma.distancia
  end

  def distancia_ate(this, outro_personagem) do
    cateto_a = :math.pow this.x - outro_personagem.x, 2
    cateto_b = :math.pow this.y - outro_personagem.y, 2
    :math.sqrt cateto_a + cateto_b
  end
end

File: arena-rpg.java/src/Personagem.java
import java.lang.Math;  

public class Personagem {

    private String nome;
    private Arma arma;
    private int x = 0, y = 0, hp;
    
    public Personagem(String nome, Arma arma, int hp) {
        this.nome = nome;
        this.arma = arma;
        this.hp = hp;
    }

    public int getHp() {
        return hp;
    }

    public void setHp(int hp) {
        this.hp = hp;
    }

    public String getNome() {
        return nome;
    }

    public Arma getArma() {
        return arma;
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public void movePara(int x, int y) {
        this.x = x;
        this.y = y;
    }
    
    public boolean estaVivo() {
        return hp > 0;
    }

    public boolean estaArmado() {
        return arma != null;
    }
    
    public int distanciaAte(Personagem outroPersonagem) {
        int a = (int) Math.pow(x - outroPersonagem.x, 2);
        int b = (int) Math.pow(y - outroPersonagem.y, 2);
        return (int) Math.sqrt(a + b);
    }
    
    public boolean estaAtingivel(Personagem outroPersonagem) {
        return estaArmado() ? (distanciaAte(outroPersonagem) <= arma.getDistancia()) : false;
    }
    
    public void ataca(Personagem outroPersonagem) {
        try {
            if(estaAtingivel(outroPersonagem)) {
                outroPersonagem.setHp(outroPersonagem.getHp() - this.arma.getDano());
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
